In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns

/tmp/ipykernel_29882/1099235115.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
df = pd.read_csv("./data/diabetes_012_health_indicators_BRFSS2021.csv")
df

Diabetes_012  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                0.0       0       1.0          1  15.0     1.0     0.0   
1                2.0       1       0.0          1  28.0     0.0     0.0   
2                2.0       1       1.0          1  33.0     0.0     0.0   
3                2.0       0       1.0          1  29.0     0.0     1.0   
4                0.0       0       0.0          1  24.0     1.0     0.0   
...              ...     ...       ...        ...   ...     ...     ...   
236373           2.0       1       1.0          1  21.0     0.0     0.0   
236374           0.0       1       0.0          1  25.0     1.0     0.0   
236375           0.0       0       1.0          1  31.0     0.0     0.0   
236376           0.0       1       0.0          1  24.0     0.0     0.0   
236377           0.0       0       1.0          1  32.0     0.0     0.0   

        HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                        0.0             0       1  ...              1   
1                        1.0             0       1  ...              1   
2                        0.0             1       1  ...              1   
3                        1.0             1       1  ...              1   
4                        0.0             0       0  ...              1   
...                      ...           ...     ...  ...            ...   
236373                   0.0             1       1  ...              1   
236374                   0.0             1       1  ...              0   
236375                   0.0             1       1  ...              1   
236376                   0.0             1       1  ...              1   
236377                   0.0             1       0  ...              1   

        NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex  Age  \
0               0.0      5.0      10.0      20.0       0.0    0   11   
1               0.0      2.0       0.0       0.0       0.0    0   11   
2               0.0      2.0      10.0       0.0       0.0    0    9   
3               0.0      5.0       0.0      30.0       1.0    1   12   
4               0.0      3.0       0.0       0.0       1.0    1   13   
...             ...      ...       ...       ...       ...  ...  ...   
236373          0.0      4.0       0.0       0.0       0.0    1   10   
236374          1.0      2.0      20.0       0.0       0.0    0    3   
236375          0.0      2.0       0.0       0.0       0.0    1    7   
236376          0.0      2.0       0.0       0.0       0.0    1   10   
236377          0.0      1.0       2.0       2.0       0.0    0    6   

        Education  Income  
0             4.0     5.0  
1             4.0     3.0  
2             4.0     7.0  
3             3.0     4.0  
4             5.0     6.0  
...           ...     ...  
236373        2.0     3.0  
236374        4.0     5.0  
236375        6.0    10.0  
236376        4.0     6.0  
236377        6.0     6.0  

[236378 rows x 22 columns]

In [3]:
df.columns

Index(['Diabetes_012', 'HighBP', 'HighChol', 'CholCheck', 'BMI', 'Smoker',
       'Stroke', 'HeartDiseaseorAttack', 'PhysActivity', 'Fruits', 'Veggies',
       'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost', 'GenHlth',
       'MentHlth', 'PhysHlth', 'DiffWalk', 'Sex', 'Age', 'Education',
       'Income'],
      dtype='object')

In [4]:
df.isna().sum()

Diabetes_012            0
HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbcCost             0
GenHlth                 0
MentHlth                0
PhysHlth                0
DiffWalk                0
Sex                     0
Age                     0
Education               0
Income                  0
dtype: int64

In [5]:


df_encoded = pd.get_dummies(df, columns=['GenHlth', 'Education', 'Age', 'Income'], 
                            prefix=['GenHlth', 'Edu', 'Age', 'Inc'], dtype=int)

for column in df_encoded.columns:
    if df_encoded[column].dtype == 'bool':
        df_encoded[column] = df_encoded[column].astype(int)


for column_name in df_encoded.columns:
   print(f"\n\n{column_name}")
   print(df_encoded[column_name].unique())

df_encoded.head




Diabetes_012
[0. 2. 1.]


HighBP
[0 1]


HighChol
[1. 0.]


CholCheck
[1 0]


BMI
[15. 28. 33. 29. 24. 40. 27. 30. 36. 31. 35. 22. 26. 45. 23. 25. 38. 21.
 32. 20. 34. 43. 18. 44. 19. 39. 37. 41. 16. 42. 50. 51. 49. 17. 52. 47.
 46. 56. 57. 48. 58. 61. 53. 63. 54. 68. 55. 62. 13. 59. 66. 77. 60. 87.
 64. 69. 72. 14. 75. 67. 71. 65. 82. 86. 70. 78. 12. 98. 84. 73. 74. 83.
 79. 88. 81. 76. 89. 90. 95. 85. 99. 91. 94. 92.]


Smoker
[1. 0.]


Stroke
[0. 1.]


HeartDiseaseorAttack
[0. 1.]


PhysActivity
[0 1]


Fruits
[1 0]


Veggies
[1 0]


HvyAlcoholConsump
[0 1]


AnyHealthcare
[1 0]


NoDocbcCost
[0. 1.]


MentHlth
[10.  0.  5. 25.  2.  7. 30.  3. 14. 20.  8.  1. 15.  4. 28. 21. 12.  6.
 22. 27. 18. 24. 13. 17. 16.  9. 19. 29. 23. 11. 26.]


PhysHlth
[20.  0. 30. 25.  4. 10.  2.  3. 15.  8. 13. 14.  1.  5.  7.  6. 18.  9.
 16. 26. 28. 21. 27. 12. 29. 11. 17. 19. 24. 23. 22.]


DiffWalk
[0. 1.]


Sex
[0 1]


GenHlth_1.0
[0 1]


GenHlth_2.0
[0 1]


GenHlth_3.0
[0 1]


GenHlth_4.0
[0 1]


<bound method NDFrame.head of         Diabetes_012  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                0.0       0       1.0          1  15.0     1.0     0.0   
1                2.0       1       0.0          1  28.0     0.0     0.0   
2                2.0       1       1.0          1  33.0     0.0     0.0   
3                2.0       0       1.0          1  29.0     0.0     1.0   
4                0.0       0       0.0          1  24.0     1.0     0.0   
...              ...     ...       ...        ...   ...     ...     ...   
236373           2.0       1       1.0          1  21.0     0.0     0.0   
236374           0.0       1       0.0          1  25.0     1.0     0.0   
236375           0.0       0       1.0          1  31.0     0.0     0.0   
236376           0.0       1       0.0          1  24.0     0.0     0.0   
236377           0.0       0       1.0          1  32.0     0.0     0.0   

        HeartDiseaseorAttack  PhysActivity  Fruits  ...  Inc_2.0  Inc

In [6]:
from sklearn.model_selection import train_test_split
X = df.drop('Diabetes_012', axis=1)  
y = df['Diabetes_012']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [8]:
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, roc_auc_score

clf = MLPClassifier(verbose=True)

clf.fit(X_train.values, y_train.values)


y_preds = clf.predict(X_test.values)

accuracy_score(y_test, y_preds)

Iteration 1, loss = 0.49826364
Iteration 2, loss = 0.43781322


/home/k4czyns/learning/webelight/.env/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:698: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


0.8395380319824012

In [ ]:
from sklearn.metrics import classification_report
report = classification_report(y_test, y_preds)
print(report)

              precision    recall  f1-score   support

         0.0       0.86      0.97      0.91     39438
         1.0       0.00      0.00      0.00      1124
         2.0       0.54      0.21      0.31      6714

    accuracy                           0.84     47276
   macro avg       0.47      0.40      0.41     47276
weighted avg       0.79      0.84      0.80     47276



In [ ]:
import pickle

# save
with open('Diabetes012MLPmodel.pkl','wb') as f:
    pickle.dump(clf,f)


In [8]:
from sklearn.metrics import accuracy_score, classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC

model_performance = []

def train_evaluate_model(model, model_name):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    print(f"{model_name} Accuracy: {accuracy:.4f}")
    print(classification_report(y_test, y_pred))
    model_performance.append((model_name, accuracy))



In [11]:
train_evaluate_model(LogisticRegression(max_iter=1000, verbose=True), "Logistic Regression")



RUNNING THE L-BFGS-B CODE

           * * *

Machine precision = 2.220D-16
 N =           66     M =           10

At X0         0 variables are exactly at the bounds

At iterate    0    f=  1.09861D+00    |proj g|=  1.39825D+01


 This problem is unconstrained.



At iterate   50    f=  4.49032D-01    |proj g|=  2.26974D-02

At iterate  100    f=  4.39409D-01    |proj g|=  1.84738D-02

At iterate  150    f=  4.32541D-01    |proj g|=  2.27425D-02

At iterate  200    f=  4.28643D-01    |proj g|=  1.41664D-02

At iterate  250    f=  4.27947D-01    |proj g|=  3.96929D-03

At iterate  300    f=  4.27725D-01    |proj g|=  2.64856D-03

At iterate  350    f=  4.27683D-01    |proj g|=  9.27728D-04

At iterate  400    f=  4.27633D-01    |proj g|=  1.44580D-03

At iterate  450    f=  4.27612D-01    |proj g|=  3.63036D-03

At iterate  500    f=  4.27584D-01    |proj g|=  1.59306D-03

At iterate  550    f=  4.27569D-01    |proj g|=  2.56490D-03

At iterate  600    f=  4.27543D-01    |proj g|=  6.29611D-04

At iterate  650    f=  4.27528D-01    |proj g|=  2.72233D-03

At iterate  700    f=  4.27521D-01    |proj g|=  2.15102D-03

At iterate  750    f=  4.27511D-01    |proj g|=  3.54550D-03

At iterate  800    f=  4.27505D-01    |proj g|=  3.45162D-03

At iter

/home/k4czyns/learning/webelight/.env/lib/python3.12/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/home/k4czyns/learning/webelight/.env/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/home/k4czyns/learning/webelight/.env/lib/python3.12/site-packages/sklearn/metrics/_classification.py:1497: UndefinedM

In [12]:
train_evaluate_model(RandomForestClassifier(verbose=True), "Random Forest")




[Parallel(n_jobs=1)]: Done  49 tasks      | elapsed:   13.0s
[Parallel(n_jobs=1)]: Done  49 tasks      | elapsed:    0.8s


Random Forest Accuracy: 0.8315
              precision    recall  f1-score   support

         0.0       0.86      0.96      0.91     39438
         1.0       0.02      0.00      0.00      1124
         2.0       0.46      0.19      0.27      6714

    accuracy                           0.83     47276
   macro avg       0.45      0.39      0.39     47276
weighted avg       0.78      0.83      0.79     47276



In [13]:
train_evaluate_model(GradientBoostingClassifier(verbose=True), "Gradient Boosting")


      Iter       Train Loss   Remaining Time 
         1           0.5010            1.43m
         2           0.4891            1.46m
         3           0.4802            1.49m
         4           0.4730            1.43m
         5           0.4671            1.41m
         6           0.4617            1.38m
         7           0.4574            1.40m
         8           0.4537            1.37m
         9           0.4503            1.34m
        10           0.4474            1.32m
        20           0.4325            1.16m
        30           0.4266            1.11m
        40           0.4238           57.31s
        50           0.4224           48.46s
        60           0.4213           37.96s
        70           0.4207           28.05s
        80           0.4201           19.02s
        90           0.4198            9.48s
       100           0.4194            0.00s
Gradient Boosting Accuracy: 0.8419
              precision    recall  f1-score   support

         

This results even-though show 84% accuracy but it can be misleading. I may try SMOTE sampling, some more hyperparameter tuning. Not currently the scope of this notebook. However might return just for the funzies.